elasticdl/python/common/data_reader.py

```diff
@@ -61,3 +64,45 @@ def create_shards(self):
             with closing(recordio.Index(p)) as rio:
                 f_records[p] = (start_ind, rio.num_records())
         return f_records
+
+
+class ODPSDataReader(AbstractDataReader):
+    def __init__(self, **kwargs):
+        AbstractDataReader.__init__(self, **kwargs)
+        self._kwargs = kwargs
+        self._reader = ODPSReader(
+            project=self._kwargs["project"],
+            access_id=self._kwargs["access_id"],
+            access_key=self._kwargs["access_key"],
+            endpoint=self._kwargs["endpoint"],
+            table=self._kwargs["table"],
+            num_processes=self._kwargs["num_processes"],
+        )
+
+    def read_records(self, task):
+        records = self._reader.read_batch(
+            start=task.start, end=task.end, columns=None
+        )
+        for batch in records:
+            yield batch
+
+    def create_shards(self):
+        shard_name_prefix = "shard_"
+        table_size = self._reader.get_table_size()
+        records_per_task = self._kwargs["records_per_task"]
+        shards = {}
+        num_shards = math.floor(table_size / records_per_task)
+        start_ind = 0
+        for shard_id in range(num_shards):
+            shards[shard_name_prefix + str(shard_id)] = (
+                start_ind,
+                records_per_task,
+            )
+            start_ind += records_per_task
+        num_records_left = table_size % records_per_task
+        if num_records_left != 0:
+            shards[shard_name_prefix + str(num_shards)] = (
+                start_ind,
+                num_records_left,
+            )
+        return shards
```

Comment:
Be careful that the size of a table can be a very large number. So `shards` may occupy a large part of the total memory.  It would be better that `shards` can be a generator.

Reply:
Good call. This could be done separately as it involves some changes in other places such as TaskDispatcher, etc.